app/models/partner.rb

```diff
@@ -311,10 +311,43 @@ def owned_neighbourhood_ids
     neighbourhood_ids
   end
 
+  def neighbourhood_name_for_site(badge_zoom_level)
+    if service_areas.any?
+      if service_areas.count > 1
+        'Various'
+      else
+        service_areas.first&.neighbourhood&.shortname
+      end
+    else
+      address&.neighbourhood&.name_from_badge_zoom(badge_zoom_level)
+    end
+  end
+
   def self.fuzzy_find_by_location(components)
     Partner.find_by('lower(name) IN (?)', components.map(&:downcase))
   end
 
+  def self.neighbourhood_names_for_site(current_site, badge_zoom_level)
+    partners = Partner.for_site(current_site)
+                      .includes(:service_areas, :address)
+    partner_names = []
+    all_partner_neighbourhoods =
+      partners.each do |partner|
+        name = partner.neighbourhood_name_for_site(badge_zoom_level)
+        partner_names << name if name
+      end
+    partner_names.uniq.sort
+  end
+
+  def self.for_neighbourhood_name_filter(query = Partner, badge_zoom_level, neighbourhood_name)
+    partners_with_name = []
+    all_partner_neighbourhoods =
```

Comment:
I ended up sticking to each here but only ever passing in Partner.all, and removing the redundant return values. Otherwise every test was breaking because for whatever 'unsolvable today' reason when you get a partner back from Partner.all in the tests, it does not come with parents attached.